transformer_engine/pytorch/attention.py

```diff
@@ -95,39 +96,97 @@
 _formatter = logging.Formatter("[%(levelname)-8s | %(name)-19s]: %(message)s")
 _stream_handler = logging.StreamHandler()
 _stream_handler.setFormatter(_formatter)
+fa_logger = logging.getLogger()
+fa_logger.setLevel(_log_level)
+if not fa_logger.hasHandlers():
+    fa_logger.addHandler(_stream_handler)
+
+
+@functools.lru_cache(maxsize=None)
+def _get_supported_versions(version_min, version_not_supported, version_max):
+    return (
+        ">= "
+        + str(version_min)
+        + ", "
+        + "".join(["!= " + str(x) + ", " if x > version_min else "" for x in version_not_supported])
+        + "<= "
+        + str(version_max)
+    )
+
 
 _NVTE_FLASH_ATTN = int(os.getenv("NVTE_FLASH_ATTN", "1"))
 _NVTE_FUSED_ATTN = int(os.getenv("NVTE_FUSED_ATTN", "1"))
 _NVTE_UNFUSED_ATTN = int(os.getenv("NVTE_UNFUSED_ATTN", "1"))
-_flash_attn_version = PkgVersion(get_pkg_version("flash-attn"))
+
+# Detect flash-attn v2 in the environment
+_flash_attn_is_installed = False
+_flash_attn_version = PkgVersion("0")
 _flash_attn_version_required = PkgVersion("2.0.6")
+_flash_attn_version_not_supported = (PkgVersion("2.0.9"), PkgVersion("2.1.0"))
```

Comment:
I think this is leading to a lot of additional maintenance. Now that we're making FA optional I suggest we move the minimum FA to 2.1.1 and get rid of this logic unless we're committed to supporting 2.0.6 - 2.0.8 (which I don't think is necessary) .

Reply:
I think the difference between pre-2.1 and post-2.1 is the causal/causal_bottom_right mask. Before 2.1, FA only supports top-left causal, and after 2.1, only bottom-right causal. I wonder if there are still users using that.

https://github.com/cyanguwa/TransformerEngine/blob/6fd94c74334800d0488a89d4cf836d96edbf0f1b/transformer_engine/pytorch/attention.py#L643

Reply:
Should we mention the version of FA we've tested with in the docs somewhere? Related issue: https://github.com/NVIDIA/TransformerEngine/issues/1236

Reply:
What do you think of [c2b3be8](https://github.com/NVIDIA/TransformerEngine/pull/1226/commits/c2b3be843aa3bd37a380da2e5446dfb5fce69a39)?

Reply:
I think this test is too strong, the test suite would grow exponentially especially if we checked each specific version of our dependencies